blog/2025-11-12-best-practises-for-writing-promises/index.mdx

```diff
@@ -0,0 +1,559 @@
+```
+
+This will have updated the `promise.yaml` to add a `resource.configure` workflow and have bootstrapped a Docker image that can be
+built to run as part of this workflow. As we has specified that we want to use the [kratix-go sdk](https://pkg.go.dev/github.com/syntasso/kratix-go) for this Promise via the `--language`
+flag, we also have a `pipeline.go` file. Update the script to contain the following:
+
+```go
+package main
+
+import (
+	"fmt"
+
+	kratix "github.com/syntasso/kratix-go"
+)
+
+func main() {
+	sdk := kratix.New()
+	resource, _ := sdk.ReadResourceInput()
+	name := resource.GetName()
+	size, err := resource.GetValue("spec.size")
+	if err != nil {
+		log.Fatalf("failed fetch 'size' property from request: %v", err)
+	}
+	
+	var replicas int
+	if size == "small" {
+		replicas = 1
+	}
+	
+	if size == "large" {
+		replicas = 3
+	}
+	
+	failoverTemplate := `apiVersion: databases.spotahome.com/v1
+kind: RedisFailover
+metadata:
+  name: %s
+  namespace: default
+spec:
+  redis:
+    imagePullPolicy: IfNotPresent
+    replicas: %d
+    customConfig:
+      - "maxclients 100"
+      - "hz 50"
+      - "timeout 60"
+      - "tcp-keepalive 60"
+      - "client-output-buffer-limit normal 0 0 0"
+      - "client-output-buffer-limit slave 1000000000 1000000000 0"
+      - "client-output-buffer-limit pubsub 33554432 8388608 60"
+    resources:
+      limits:
+        cpu: 400m
+        memory: 500Mi
+      requests:
+        cpu: 100m
+        memory: 100Mi
+  sentinel:
+    imagePullPolicy: IfNotPresent
+    replicas: %d
+    customConfig:
+      - "down-after-milliseconds 2000"
+      - "failover-timeout 3000"
+    resources:
+      limits:
+        memory: 100Mi
+      requests:
+        cpu: 100m
+`
+		redisContent := []byte(fmt.Sprintf(failoverTemplate, name, replicas, replicas))
+	
+		sdk.WriteOutput("redis_failover.yaml", redisContent)
+}
+```
+
+We’re going to execute this script when creating and updating a resource request. It will:
+
+1. Retrieve the name and size of the Redis from the resource request
+2. Interpolate these in the template for the generated redis failover alongside the configured defaults
+3. Write the RedisFailover CR to the `/kratix/output` directory so it can be written to a destination.
+
+As this workflow is written declaratively, there will be no attempts to create the Redis after initial creation. It will only be
+updated if there is a change in the specification of the resource.
+
+**Idempotency**
+
+Idempotency refers to the ability to run an action multiple times with the guarantee that there will be no additional effect of
+running after it the initial action. In our case, if the resource configure workflow runs a second or third time, no additional
+Redis instance will be created on these occasions. This is part of what having a declarative workflow enables. As the workflow
+describes the *desired* Redis instance, if the workflow re-runs for any reason, the Redis will not be created again. It will
```

Comment:
```suggestion
describes the *desired* Redis instance, if the workflow re-runs for any reason, this instance will not be created again. It will, 
```

Almost sure we will need a comma before however.